cpp/include/raft/linalg/detail/cublaslt_wrappers.hpp

```diff
@@ -0,0 +1,294 @@
+    int algo_count;
+    cublasLtMatmulPreference_t preference;
+    RAFT_CUBLAS_TRY(cublasLtMatmulPreferenceCreate(&preference));
+    RAFT_CUBLAS_TRY(cublasLtMatmulAlgoGetHeuristic(resource::get_cublaslt_handle(res),
+                                                   r.desc,
+                                                   r.a,
+                                                   r.b,
+                                                   r.c,
+                                                   r.c,
+                                                   preference,
+                                                   1,
+                                                   &r.heuristics,
+                                                   &algo_count));
+    RAFT_CUBLAS_TRY(cublasLtMatmulPreferenceDestroy(preference));
+    return r;
+  }
+};
+
+/** Cache with the default constructor; tagged with input types to use separate caches. */
+template <typename S, typename A, typename B, typename C, bool DevicePointerMode>
+struct matmul_cache {
+  /** Number of matmul invocations to cache. */
+  static constexpr size_t kDefaultSize = 100;
+  cache::lru<matmul_key_t, matmul_key_hash, std::equal_to<>, std::shared_ptr<matmul_desc>> value{
+    kDefaultSize};
+};
+
+/**
+ * @brief the wrapper of cublasLt matmul function
+ *  It computes the following equation: C = alpha .* opA(A) * opB(B) + beta .* C
+ *
+ * @tparam DevicePointerMode whether pointers alpha, beta point to device memory
+ * @tparam S the type of scale parameters alpha, beta
+ * @tparam A the element type of matrix A
+ * @tparam B the element type of matrix B
+ * @tparam C the element type of matrix C
+ *
+ * @param [in] res raft resources
+ * @param [in] trans_a cublas transpose op for A
+ * @param [in] trans_b cublas transpose op for B
+ * @param [in] m number of rows of C
+ * @param [in] n number of columns of C
+ * @param [in] k number of rows of opB(B) / number of columns of opA(A)
+ * @param [in] alpha host or device scalar
+ * @param [in] a_ptr such a matrix that the shape of column-major opA(A) is [m, k]
+ * @param [in] lda leading dimension of A
+ * @param [in] b_ptr such a matrix that the shape of column-major opA(B) is [k, n]
+ * @param [in] ldb leading dimension of B
+ * @param [in] beta host or device scalar
+ * @param [inout] c_ptr column-major matrix of size [m, n]
+ * @param [in] ldc leading dimension of C
+ * @param [in] stream
+ */
+template <bool DevicePointerMode = false, typename S, typename A, typename B, typename C>
+void matmul(raft::resources const& res,
+            bool trans_a,
+            bool trans_b,
+            uint64_t m,
+            uint64_t n,
+            uint64_t k,
+            const S* alpha,
+            const A* a_ptr,
+            uint64_t lda,
+            const B* b_ptr,
+            uint64_t ldb,
+            const S* beta,
+            C* c_ptr,
+            uint64_t ldc,
+            cudaStream_t stream)
```

Comment:
This should either accept cublaslt handle + stream or raft::resources but not raft::resources + stream. This is too error-prone and confusing for potential users.

Reply:
I totally agree we need to do this, but I'm a little bit hesitant to do that in this PR. 
  1. Removing the resources argument is definitely not an option. In cublaslt_wrappers `matmul`, we do need the resources handle, because we get the cache store from it as well, which is better not exposed in the arguments to make sure all users use the same cache.
  2. Removing the stream is doable but tricky. The problem is that both resources handle and the stream are present in all but one (which is not used anywhere) overload of the public gemm in `raft/linalg/gemm.cuh`. If we are to remove the stream in the detail namespace, we need to create a new resources handle in the public functions on each invocation to set it. I'm afraid this can add a bit of unwanted overhead, especially in the solvers like lstsq or rsvd, where gemms are invoked in a tight loop.

Would you like me to proceed with option 2?

Reply:
Another option (3). I've noticed our public gemm still has the `.cuh` extension. I suggest we can do the following:
  - Keep `linalg/detail/cublaslt_wrappers.hpp` as-is in the PR for now.
  - Add couple new overloads to `linalg/detail/gemm.hpp` that do not have the stream argument.
  - Deprecate the whole header `linalg/gemm.cuh`.
  - Add a new public header `linalg/gemm.hpp` with all overloads not having the stream argument.
  - In the future follow-ups switch to using the new gemm header and only update the `linalg/detail/cublaslt_wrappers.hpp` after we remove the deprecated  `linalg/gemm.cuh`.

Reply:
I'm okay with this third option, and have also been suggesting to folks that we rename the `cuh` extensions to `hpp` as they've been popping up in PR diffs (spgemm in @rhdong's PR as a recent example). 

Though, I'd even go as far as to say that we should deprecate `linalg/gemm.cuh` and only add the mdspan-based functions to `linalg.hpp`, since those are the "official" public API vocabulary going forward.

Reply:
Sounds good. I added new `linalg/gemm.hpp` for now with the mdspan api only and deprecated the rest.